GoRun / Test / Debug not working with multi modules directories

### What did you expect to happen?
I have a single repository with many go modules each in its own respective directory.
I usually open nvim under the root of the repository and working on the different modules.

Code completion and other features with gopls work properly on any file I work on while my working directory is the root of repository.

However, when I try and run / test / debug one of the go files I work on I get several errors in the following format:
`cannot find module providing package github.com/XXX: working directory is not part of a module`

This is also reproducible when I try and do go run with the full file path while I'm in the root folder.
Is there a way to overcome this ? 

I tried working with autochdir but that starts to break many plugins I like to use like NERDTree and FZF

### Configuration (**MUST** fill this out):

#### vim-go version:

####  `vimrc` you used to reproduce:
<!--
Use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc.

If this is not provided or is obviously incomplete, the issue may be unceremoniously closed.
-->
<!-- vimrc -->
<details><summary>vimrc</summary><br><pre>

</pre></details>

#### Vim version (first three lines from `:version`):
<!-- :version -->
NVIM v0.4.3
#### Go version (`go version`):
<!-- go version -->
go version go1.14 darwin/amd64

#### Go environment
<details><summary><code>go env</code> Output:


</summary><br><pre>
<!-- go env -->
GO111MODULE=""
GOARCH="amd64"
GOBIN="/Users/yarin/go/bin"
GOCACHE="/Users/yarin/Library/Caches/go-build"
GOENV="/Users/yarin/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/yarin/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.14/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.14/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/_d/xzxqwkds3cxb18cm25lwgqch0000gn/T/go-build228462867=/tmp/go-build -gno-record-gcc-switches -fno-common"

</pre></details>

#### gopls version
<details><summary><code>gopls version</code> Output:</summary><br><pre>
<!-- gopls version -->
golang.org/x/tools/gopls v0.3.3
    golang.org/x/tools/gopls@v0.3.3 h1:mTFqRDJQmpSsgDDWvbtGnSva1z9uX2XcDszSWa6DhBQ=
</pre></details>

https://github.com/yarinm/vim-go-issue

It reproduces very easily in my example:
Notice I created go.mod in the root

1. Open VIM from the root directory and open mycomp/main.go, from there do ":GoRun" and it works
2. Remove the go.mod from the root directory
3. Reopen VIM from the root directory, open mycomp/main.go and run :GoRun again

I currently get:
```
:!go run '/Users/yarin/workspace/vim-go-issue/mycomp/main.go'  2>&1| tee /var/folders/bv/_kllxtj105v9nhkbt5r2rslh0000gn/T/nvimNGNwlu/4
mycomp/main.go:3:8: cannot find package "yarin/mylib" in any of:
        /usr/local/Cellar/go/1.14/libexec/src/yarin/mylib (from $GOROOT)
        /Users/yarin/go/src/yarin/mylib (from $GOPATH)
```

I'm trying to run (when I don't have the go.mod file):
```
:GoAddWorkspace mycomp/
:GoAddWorkspace mylib/
```

And run the :GoRun again but I get the same error..

--- a/autoload/go/cmd.vim
+++ b/autoload/go/cmd.vim
@@ -154,10 +154,10 @@ function! go#cmd#Run(bang, ...) abort
 
   call go#statusline#Update(expand('%:p:h'), l:status)
 
-  let cmd = "go run "
-  let tags = go#config#BuildTags()
-  if len(tags) > 0
-    let cmd .= "-tags " . go#util#Shellescape(tags) . " "
+  let l:cmd = "go run "
+  let l:tags = go#config#BuildTags()
+  if len(l:tags) > 0
+    let l:cmd .= "-tags " . go#util#Shellescape(l:tags) . " "
   endif
 
   if a:0 == 0
@@ -166,17 +166,17 @@ function! go#cmd#Run(bang, ...) abort
     let l:files = map(copy(a:000), "expand(v:val)")
   endif
 
-  let cmd = printf('%s%s', cmd, go#util#Shelljoin(l:files, 1))
+  let l:cmd = printf('%s%s', l:cmd, go#util#Shelljoin(l:files, 1))
 
-  let cd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
-  let dir = getcwd()
+  let l:cd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
+  let l:dir = getcwd()
 
   if go#util#IsWin()
     try
-      execute cd . fnameescape(expand("%:p:h"))
-      exec printf('!%s', cmd)
+      execute l:cd . fnameescape(expand("%:p:h"))
+      exec printf('!%s', l:cmd)
     finally
-      execute cd . fnameescape(dir)
+      execute l:cd . fnameescape(l:dir)
     endtry
 
     let l:status.state = 'success'
@@ -198,8 +198,8 @@ function! go#cmd#Run(bang, ...) abort
   endif
 
   " :make expands '%' and '#' wildcards, so they must also be escaped
-  let default_makeprg = &makeprg
-  let &makeprg = cmd
+  let l:default_makeprg = &makeprg
+  let &makeprg = l:cmd
 
   let l:listtype = go#list#Type("GoRun")
 
@@ -209,7 +209,7 @@ function! go#cmd#Run(bang, ...) abort
     " backup user's errorformat, will be restored once we are finished
     let l:old_errorformat = &errorformat
     let &errorformat = s:runerrorformat()
-    execute cd . fnameescape(expand("%:p:h"))
+    execute l:cd . fnameescape(expand("%:p:h"))
     if l:listtype == "locationlist"
       exe 'lmake!'
     else
@@ -217,9 +217,9 @@ function! go#cmd#Run(bang, ...) abort
     endif
   finally
     "restore the working directory, errformat, and makeprg
-    execute cd . fnameescape(dir)
+    execute cd . fnameescape(l:dir)
     let &errorformat = l:old_errorformat
-    let &makeprg = default_makeprg
+    let &makeprg = l:default_makeprg
   endtry
 
   let l:errors = go#list#Get(l:listtype)

--- a/autoload/go/cmd.vim
+++ b/autoload/go/cmd.vim
@@ -160,12 +160,24 @@ function! go#cmd#Run(bang, ...) abort
     let cmd .= "-tags " . go#util#Shellescape(tags) . " "
   endif
 
+  if a:0 == 0
+    let l:files = go#tool#Files()
+  else
+    let l:files = map(copy(a:000), "expand(v:val)")
+  endif
+
+  let cmd = printf('%s%s', cmd, go#util#Shelljoin(l:files, 1))
+
+  let cd = exists('*haslocaldir') && haslocaldir() ? 'lcd ' : 'cd '
+  let dir = getcwd()
+
   if go#util#IsWin()
-    if a:0 == 0
-      exec '!' . cmd . go#util#Shelljoin(go#tool#Files(), 1)
-    else
-      exec '!' . cmd . go#util#Shelljoin(map(copy(a:000), "expand(v:val)"), 1)
-    endif
+    try
+      execute cd . fnameescape(expand("%:p:h"))
+      exec printf('!%s', cmd)
+    finally
+      execute cd . fnameescape(dir)
+    endtry
 
     let l:status.state = 'success'
     if v:shell_error
@@ -187,26 +199,25 @@ function! go#cmd#Run(bang, ...) abort
 
   " :make expands '%' and '#' wildcards, so they must also be escaped
   let default_makeprg = &makeprg
-  if a:0 == 0
-    let &makeprg = cmd . go#util#Shelljoin(go#tool#Files(), 1)
-  else
-    let &makeprg = cmd . go#util#Shelljoin(map(copy(a:000), "expand(v:val)"), 1)
-  endif
+  let &makeprg = cmd
 
   let l:listtype = go#list#Type("GoRun")
 
   let l:status.state = 'success'
+
   try
     " backup user's errorformat, will be restored once we are finished
     let l:old_errorformat = &errorformat
     let &errorformat = s:runerrorformat()
+    execute cd . fnameescape(expand("%:p:h"))
     if l:listtype == "locationlist"
       exe 'lmake!'
     else
       exe 'make!'
     endif
   finally
-    "restore errorformat
+    "restore the working directory, errformat, and makeprg
+    execute cd . fnameescape(dir)
     let &errorformat = l:old_errorformat
     let &makeprg = default_makeprg
   endtry

